Add Dockerfile(s)

enkore opened the pull request:
base dockerfile -> https://hub.docker.com/r/qabel/base/

Contains dependencies, no application. Basically a dep pin.

app dockerfile -> https://hub.docker.com/r/qabel/infrastructure/

Contains the apps.

The inv tasks in the docker.XYZ namespace are administrative in nature (although `inv docker.run` _does_ save a lot of typing, since it sets up the port configuration etc. -- the vagrant stuff in core basically does the same thing)

Fixes #1

enkore committed:
Add Dockerfile(s)

enkore committed:
Don't change behaviour.

enkore committed:
Deploy in parallel (EXPERIMENTAL)

enkore committed:
ref: APPS_AND_DEPLOY_TASKS -> APPS

enkore committed:
Cleanup dockerignore

enkore committed:
Animated status output.

enkore committed:
quell redis/pgsql log noise

enkore committed:
Neater error handling, serialize tasks per app \o/

enkore committed:
inv update: don't pull right after cloning (less noisy output)

enkore commented:
I think I'm done here unless someone has wishes or finds a bug :)

julianseeger commented:
Ftr.: Time to get it running on your favorite OS (without containers): couple of weeks and counting...

enkore committed:
fixed unintentionally short timeout

julianseeger commented:
Did you check if docker-hub is able to build multiple containers from one repository? Otherwise we would need to drop the base image

julianseeger commented on docker_base/Dockerfile:
```diff
@@ -0,0 +1,18 @@
+    apt-get install -y postgresql-9.5 libpq-dev \
+                       redis-server redis-tools \
+                       build-essential libffi-dev git \
+                       python3.5 python3-virtualenv virtualenv python3-pip && \
```
Best practice (according to the docs) is to sort the arguments alphanumerically and to put every package in a new line. I understand the blocks for python, postgres and redis but you could split the build tools (because they are probably the first to get extended)

julianseeger commented on docker_base/Dockerfile:
```diff
@@ -0,0 +1,18 @@
+                       build-essential libffi-dev git \
+                       python3.5 python3-virtualenv virtualenv python3-pip && \
+    apt-get clean && \
+    pip3 install -U pip setuptools wheel
```
there will be a pip cache leftover won't it?

enkore replied:
1.67 MB. I can move the pip-cache-disable from the app dockerfile here if you want.

julianseeger replied:
won't hurt, especially since it's already there

julianseeger commented on docker_base/Dockerfile:
```diff
@@ -0,0 +1,18 @@
+# This is a base container we use. It has pgsql, redis and Python 3.5,
+# plus a user named qabel and locales.
+
+FROM ubuntu:16.04
```
can't we get a smaller base image? How big is the resulting image and how much of that is consumed by the base image?
edit: base image is about 134MB ... and arguing with "the layers are cached" probably won't work for long because they are updated too and we should pull updates in

enkore replied:
@rehwanne suggested alpine linux (2 MB), but almost nothing we need (postgres etc.) is packaged for that. Given that Debian stable has no Python 3.5 and stretch is rolling-release, and that we run on Ubuntu in production made it the logical choice. Going from Debian to Ubuntu saved some ~100 MB. If there's a smaller base image around where I don't have to start packaging every dependency I'll gladly use that instead.

The qabel/base image is ~250 MB, the app image ~400 MB (the pgsql DB is ~70 MB - many small files, maybe tunable?). The rest are the dependencies, there's quite a bit of data in there. Seems to compress fairly well.

julianseeger replied:
Ok sounds reasonable. As the "normal" use would be to just link a postgres container nobody complies about alpine linux not support it ^^

enkore replied:
Also see the comment below. The download is actually much smaller than the on-disk size, so it's probably not an issue.

enkore commented:
> Did you check if docker-hub is able to build multiple containers from one repository? Otherwise we would need to drop the base image

I don't know, the account linkage isn't working for me. Probably because this isn't in my account. I'll try clicking around some more in the web ui. Edit: I tried relinking, now "public and private". Pending approval. Maybe it makes a difference (even though this isn't private?).

enkore commented:
Downloads on a clean docker (no images): 50 MB, 200 MB, 11 MB, 34 MB, 141 MB = 436 MB total (and a handful of small layers)

If the layers in the docker-pull are listed as they are logically composed then that's 50 MB base image, 200 MB base deps image, 11 MB is probably the ADD step, 34 MB could be the bootstrap step and 141 MB is the deployed layer I guess.

julianseeger commented:
> I don't know, the account linkage isn't working for me. Probably because this isn't in my account. I'll try clicking around some more in the web ui. Edit: I tried relinking, now "public and private". Pending approval. Maybe it makes a difference (even though this isn't private?).

The request arrived but we shouldn't need that. With [a limited linked account](https://docs.docker.com/docker-hub/github/#/auto-builds-and-limited-linked-github-accounts) you should be able to manually trigger a build. But we probably need 1 repo per image

enkore committed:
trademark typo

enkore committed:
partial revert of da734b8

changed start there to not redeploy stuff because it's pointless
in the container. however, since it's now faster than ever and
will never take long because there's nothing to do, i revert that change.

enkore committed:
Revert "Don't change behaviour."

This reverts commit 225ed755acacc18672ee92ded21a4915cef66a16.

enkore committed:
Sort package names and split build group.

enkore committed:
Move pip --no-cache to base image => v4.

enkore committed:
Instead of pg_ctl-waiting for postgres, try doing queries

enkore commented:
merge? the automatic building probably won't require any changes here.

enkore committed:
index, enable FACET_SHALLOW_VERIFICATION

enkore committed:
index fix for core tests
